### PR TITLE
ENH: Add fft optional extension submodule to numpy.array_api

### PR DESCRIPTION
**Pull request by asmeurer**
Note: we need to address a few issues upstream in the standard and the array-api-tests before this should be merged.

### CODE REVIEW

**Comment by asmeurer** [comment]
This should be ready for review. It's possible some of the dtype restrictions will change in the spec (see the discussion at https://github.com/data-apis/array-api/pull/720#discussion_r1423281070), but if that happens we can update this with another PR.

**Comment by charris** [comment]
Will need a release note.

**Comment by jakevdp** [comment]
I think the tests are still skipped: https://github.com/numpy/numpy/blob/6e3b92339a912f5326ff434a6f32e015c27369db/tools/ci/array-api-skips.txt#L37

When trying this in JAX, I found that `fft_test.py` in the test suite has a number of dtype expectations that don't align with the dtypes of `jax.numpy.fft`, which was designed to follow `numpy.fft`. So I'm interested to see if you run into the same isssue.

**Comment by asmeurer** [comment]
I didn't realize that the array api tests are run on the numpy CI now. I've been running them manually. The dtype stuff is addressed here. The fft functions in numpy all upcast 32-bit inputs, which the spec says they shouldn't. This is addressed here by re-downcasting, but I also plan to fix this in numpy.fft itself (my understanding is that this should be straightforward once https://github.com/numpy/numpy/pull/25536 is merged). There are also some issues in the test suite itself so we may need to get those addressed before we remove the skips.

File: numpy/array_api/fft.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    axis: int = -1,
+    norm: Literal["backward", "ortho", "forward"] = "backward",
+) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.fft.fft <numpy.fft.fft>`.
+
+    See its docstring for more information.
+    """
+    if x.dtype not in _complex_floating_dtypes:
+        raise TypeError("Only complex floating-point dtypes are allowed in fft")
```
Comment (jakevdp) [inline numpy/array_api/fft.py:35]: In the 2022.12 version of the standard, it says the inputs should be floating point and outputs should be complex: https://data-apis.org/array-api/2022.12/extensions/generated/array_api.fft.fft.html

Comment (jakevdp) [reply numpy/array_api/fft.py:35]: (but I found the dtypes used in the test suite disagree with the dtypes in the documentation in many places)

Comment (asmeurer) [reply numpy/array_api/fft.py:35]: https://github.com/data-apis/array-api/pull/720

Comment (asmeurer) [reply numpy/array_api/fft.py:35]: And yes, the test suite for fft needs to be updated in many places. I'm not aware of other places outside of fft that are wrong, so if you find any please open issues for those.

**Review by rgommers — COMMENTED** [review]
Thanks Aaron, and thanks Jake for the review. I think this is close - I think I found one mistake in dtype handling still, and the `device` handling will need an update either here or in gh-25370 depending on which PR is merged first.

**Review by rgommers — APPROVED** [review]
This looks good to go now, thanks Aaron. In it goes.

**Comment by lucascolley** [comment]
I'll test this in SciPy with a NumPy nightly soon

**Comment by lucascolley** [comment]
> Note: we need to address a few issues upstream in the standard and the array-api-tests before this should be merged

Does this mean that array-api-tests is up to scratch for `fft` now and we can remove the skip in CI here?

**Comment by asmeurer** [comment]
> Does this mean that array-api-tests is up to scratch for fft now and we can remove the skip in CI here?


No, the test suite still needs to be updated for https://github.com/data-apis/array-api/pull/720, and also some of the shape tests are wrong. (cc @honno)

At any rate, as soon as https://github.com/numpy/numpy/pull/25370 is merged the plan is to move numpy.array_api to another repo, and (presumably) delete it from this one, so I wouldn't worry too much right now about the array api CI skips for numpy.array_api. As for the main namespace fft, it is still incorrect in upcasting 32-bit inputs. I plan to fix this, but I need https://github.com/numpy/numpy/pull/25536 first, as I understand it.

**Pull request merged**